ops/model.py

```diff
@@ -1326,6 +1327,145 @@ def list_files(self, path: str, *, pattern: str = None,
         """
         return self._pebble.list_files(path, pattern=pattern, itself=itself)
 
+    def push_path(self,
+                  source_path: typing.Union[typing.Union[str,
+                                                         Path],
+                                            typing.List[typing.Union[str,
+                                                                     Path]]],
+                  dest_dir: typing.Union[str,
+                                         Path]):
+        """Push all the things.
+
+        TODO: write this docstring
+        """
+        source_paths = source_path
+        if not isinstance(source_paths, list):
+            source_paths = [source_paths]
+        source_paths = [str(p) for p in source_paths]
+        dest_dir = str(dest_dir)
+
+        def local_list(source_path: str):
+            paths = [source_path]
```

Comment:
Similarly:

```python
paths = os.path.listdir(source_path) if os.path.isdir(source_path) else [source_path]
```

This doesn't have to be a ternary, but it's otherwise a little bit of a headscratcher to see "huh, we assign `paths` then overwrite it on the next line". If it's not a ternary, an `if/else` makes it read a little more naturally.

Reply:
Swapped for ternary to appease the functional gods.

Reply:
Hey don't swap to ternary just for me ;) An `if/else` instead of "assign and immediately overwrite" would be great, too.

The "big" comments are basically:

* Please move instantiation of `pebble.FileInfo` objects to a method, and hopefully one which passes args by name (`pebble.FileInfo(name=...)`) so it's a) hidden from the reader who may wonder why these 10 variables must be in that particular order, b) less dependent on an exact implementation of `FileInfo.__init__`
* `os.path.isfile()` or `{Path}.isfile()` can clean up most of the `if stat.S_FOO(...)` calls

The new type is already great :D

Reply:
This is all internal implementation.  And this particular code is only used right here - nowhere else.  I don't find the case for a separate func particularly compelling.  I can certainly go full kward style when calling the FileInfo constructor though.  The FileInfo object and its constructor are already public API.  If I can't depend on our own public API inside our private API, then I'm not sure we should be writing any code :joy:

Reply:
It is internal, but even something like `_build_destpath` would help.

You're not wrong about `FileInfo` and it constructor being public API, but 10 args is also a lot without naming them, and 10 args is a lot to encounter in the middle of a block like this when all of them can be derived by passing `path` to `_build_fileinfo` and getting a `FileInfo` back. The way it is at present, pretty much any change to the `FileInfo` constructor would require an author to go count args [here](https://github.com/canonical/operator/pull/754/files#diff-9ac334e8310ae3abff33570665575179e0ff2ffa874efb65bd773723989bd251R1373) to make sure they make the "right" change. `ops.pebble.FileInfo.from_dict` even [does this internally](https://github.com/canonical/operator/blob/6ee44cfb64d0680ec7e72ad1c746a87a0c63f0c8/ops/pebble.py#L863).

Shuffling it off, outside of looking cleaner, follows a pattern that's already sort-of there in `_build_destpath` for a helper method, and follows keyword naming [at least some args when there are a lot](https://github.com/canonical/operator/pull/754/files#diff-9ac334e8310ae3abff33570665575179e0ff2ffa874efb65bd773723989bd251R1397).

Outside of anything else, it safeguards against a possible "whoops, this broke but we didn't catch it" change, since even the [unit tests in this patch](https://github.com/canonical/operator/pull/754/files#diff-b7713b8731a4d1f895063e58f31f8ab54f7bb6ca0e2b1712d85683874d94a6e6R819) assign args by name (and dict unpack the rest the same way). There's a hypothetical scenario here where `pebble.FileInfo` _does_ change slightly, it skates through the unit tests, and throws exceptions in running code

Reply:
I took some of your suggestions and munged them with some of mine